R/prof.to.features.R

```diff
@@ -1,622 +1,706 @@
+    par(mfrow = c(ceiling(length(all.bw) / 2), 2))
+    par(mar = c(1, 1, 1, 1))
+  }
+
+  last.num.pks <- Inf
+
+  chr_profile_unfiltered <- chr_profile
+  chr_profile <- data.frame(chr_profile) %>%
+    filter(intensity > 1e-5) %>%
+    arrange(base_curve)
+
+  for (bw.n in length(all.bw):1)
+  {
+    bw <- all.bw[bw.n]
+    this.smooth <- ksmooth(chr_profile_unfiltered[, "base_curve"], chr_profile_unfiltered[, "intensity"], kernel = "normal", bandwidth = bw)
+    turns <- find.turn.point(this.smooth$y)
+    pks <- this.smooth$x[turns$pks]
+    vlys <- c(-Inf, this.smooth$x[turns$vlys], Inf)
+
+    smoother.pk.rec[[bw.n]] <- pks
+    smoother.vly.rec[[bw.n]] <- vlys
+    if (length(pks) != last.num.pks) {
+      last.num.pks <- length(pks)
+      l <- length(chr_profile[, "base_curve"])
+      dx <- c(chr_profile[2, "base_curve"] - chr_profile[1, "base_curve"], (chr_profile[3:l, "base_curve"] - chr_profile[1:(l - 2), "base_curve"]) / 2, chr_profile[l, "base_curve"] - chr_profile[l - 1, "base_curve"])
+      if (l == 2) {
+        dx <- rep(diff(chr_profile[, "base_curve"]), 2)
+      }
+
+      # initiation
+      m <- s1 <- s2 <- delta <- pks
+      for (i in 1:length(m))
+      {
+        sel.1 <- which(chr_profile[, "base_curve"] >= max(vlys[vlys < m[i]]) & chr_profile[, "base_curve"] < m[i])
+        s1[i] <- sqrt(sum((chr_profile[sel.1, "base_curve"] - m[i])^2 * chr_profile[sel.1, "intensity"] * dx[sel.1]) / sum(chr_profile[sel.1, "intensity"] * dx[sel.1]))
+
+        sel.2 <- which(chr_profile[, "base_curve"] >= m[i] & chr_profile[, "base_curve"] < min(vlys[vlys > m[i]]))
+        s2[i] <- sqrt(sum((chr_profile[sel.2, "base_curve"] - m[i])^2 * chr_profile[sel.2, "intensity"] * dx[sel.2]) / sum(chr_profile[sel.2, "intensity"] * dx[sel.2]))
+
+        delta[i] <- (sum(chr_profile[sel.1, "intensity"] * dx[sel.1]) + sum(chr_profile[sel.2, "intensity"] * dx[sel.2])) / ((sum(dnorm(chr_profile[sel.1, "base_curve"], mean = m[i], sd = s1[i])) * s1[i] / 2) + (sum(dnorm(chr_profile[sel.2, "base_curve"], mean = m[i], sd = s2[i])) * s2[i] / 2))
+      }
+      delta[is.na(delta)] <- 1e-10
+      s1[is.na(s1)] <- 1e-10
+      s2[is.na(s2)] <- 1e-10
+
+
+      fit <- matrix(0, ncol = length(m), nrow = length(chr_profile[, "base_curve"])) # this is the matrix of fitted values
+
+      this.change <- Inf
+      counter <- 0
+
+      while (this.change > 0.1 & counter <= max.iter) {
+        counter <- counter + 1
+        old.m <- m
+
+        # E step
+        cuts <- c(-Inf, m, Inf)
+        for (j in 2:length(cuts))
+        {
+          sel <- which(chr_profile[, "base_curve"] >= cuts[j - 1] & chr_profile[, "base_curve"] < cuts[j])
+          use.s1 <- which(1:length(m) >= (j - 1))
+          s.to.use <- s2
+          s.to.use[use.s1] <- s1[use.s1]
+          for (i in 1:ncol(fit))
+          {
+            fit[sel, i] <- dnorm(chr_profile[sel, "base_curve"], mean = m[i], sd = s.to.use[i]) * s.to.use[i] * delta[i]
+          }
+        }
```

Comment:
It says E step, but there is no check whether the method chosen is `EM` or `moment`?

Reply:
The check occurs after this "E" step. In the step "M" for whatever reason :\